### PR TITLE
Fix #11830

### LINKED ISSUE #11830: mingw32 build fails due to missing nix::readLine()
**Describe the bug**

https://hydra.nixos.org/build/277363338

```
[6/6] Linking target libnixmain.dll
FAILED: libnixmain.dll 
x86_64-w64-mingw32-g++  -o libnixmain.dll libnixmain.dll.p/common-args.cc.obj libnixmain.dll.p/loggers.cc.obj libnixmain.dll.p/plugin.cc.obj libnixmain.dll.p/progress-bar.cc.obj libnixmain.dll.p/shared.cc.obj -Wl,--allow-shlib-undefined -shared -Wl,--start-group -Wl,--out-implib=libnixmain.dll.a /nix/store/jdqqx9nnz5yp0myjyjrblbdk8hxk60kc-nix-util-x86_64-w64-mingw32-2.25.0pre/lib/libnixutil.dll.a /nix/store/b2n32ylfcw3dsxznj81akd8539ad4vlb-libarchive-x86_64-w64-mingw32-3.7.4-lib/lib/libarchive.dll.a /nix/store/llhkczan46l8x9v11sv7cb36p6nrkhhi-nix-store-x86_64-w64-mingw32-2.25.0pre/lib/libnixstore.dll.a -lkernel32 -luser32 -lgdi32 -lwinspool -lshell32 -lole32 -loleaut32 -luuid -lcomdlg32 -ladvapi32 -Wl,--end-group
/nix/store/4bkj4cch6qjy02254b3dwpgc7sfkk9v0-x86_64-w64-mingw32-binutils-2.41/bin/x86_64-w64-mingw32-ld: libnixmain.dll.p/progress-bar.cc.obj:progress-bar.c:(.text$_ZN3nix11ProgressBar3askESt17basic_string_viewIcSt11char_traitsIcEE[_ZN3nix11ProgressBar3askESt17basic_string_viewIcSt11char_traitsIcEE]+0x103): undefined reference to `nix::readLine[abi:cxx11](void*, bool)'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```

**Steps To Reproduce**

**Expected behavior**

**`nix-env --version` output**

**Additional context**

**Priorities**

Add :+1: to [issues you find important](https://github.com/NixOS/nix/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).

### PR DESCRIPTION
**Pull request by Ericson2314**
# Motivation

Fix #11830

# Context

See each commit for a little details.

# Priorities and Process

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

### CODE REVIEW

File: src/libutil/file-descriptor.hh [hunk 1]
```diff
@@ -106,15 +106,42 @@ void drainFD(
 #endif
     );
 
+/**
+ * Get [Standard Input](https://en.wikipedia.org/wiki/Standard_streams#Standard_input_(stdin))
+ */
+[[gnu::always_inline]]
+inline Descriptor getStandardInput() {
```
Comment (edolstra) [inline src/libutil/file-descriptor.hh:113]:
```suggestion
inline Descriptor getStandardInput()
{
```

File: src/libutil/file-descriptor.hh [hunk 2]
```diff
@@ -106,15 +106,42 @@ void drainFD(
...
+    return GetStdHandle(STD_INPUT_HANDLE);
+#endif
+}
+
+/**
+ * Get [Standard Output](https://en.wikipedia.org/wiki/Standard_streams#Standard_output_(stdout))
+ */
 [[gnu::always_inline]]
-inline Descriptor getStandardOut() {
+inline Descriptor getStandardOutput() {
```
Comment (edolstra) [inline src/libutil/file-descriptor.hh:125]:
```suggestion
inline Descriptor getStandardOutput()
{
```

File: src/libutil/file-descriptor.hh [hunk 3]
```diff
@@ -106,15 +106,42 @@ void drainFD(
...
 #else
     return GetStdHandle(STD_OUTPUT_HANDLE);
 #endif
 }
 
+/**
+ * Get [Standard Error](https://en.wikipedia.org/wiki/Standard_streams#Standard_error_(stderr))
+ */
+[[gnu::always_inline]]
+inline Descriptor getStandardError() {
```
Comment (edolstra) [inline src/libutil/file-descriptor.hh:137]:
```suggestion
inline Descriptor getStandardError()
{
```

File: src/libutil/file-descriptor.hh [hunk 4]
```diff
@@ -106,15 +106,42 @@ void drainFD(
 #endif
     );
 
+/**
+ * Get [Standard Input](https://en.wikipedia.org/wiki/Standard_streams#Standard_input_(stdin))
+ */
+[[gnu::always_inline]]
```
Comment (edolstra) [inline src/libutil/file-descriptor.hh:112]: Do we need the always_inline? Might be a bit of a micro-optimisation.

**Review by edolstra — APPROVED** [review]

File: src/libutil/file-descriptor.hh (hunk 4 above)
Comment (Ericson2314) [reply src/libutil/file-descriptor.hh:112]: I don't know. @puffnfresh did this, so I defer to him.

Comment (Ericson2314) [reply src/libutil/file-descriptor.hh:112]: (Gonna merge in the meantime since this decision is orthogonal)

**Pull request merged**